[PhpUnitBridge] Use triggering class to generate baseline for deprecation messages from DebugClassLoader

leongersen opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 5.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       | Fix #45943
| License       | MIT
| Doc PR        | N/A

Fixes #45943 by using the triggering class to generate the baseline for messages from `DebugClassLoader`. 

This would break currently generated baselines, but as #45943 describes, those are broken already when running tests in another order, or when running a sub-/superset of tests.

I'll finish this PR by adding/updating tests if this approach is acceptable.

leongersen committed:
Use triggering class to generate baseline for deprecation messages from DebugClassLoader

ruudk commented:
@leongersen Thanks for working on fixing this problem. Would you be able to add the failing test? It would be super if this problem could be patched. 

You can send me a Tikkie for a nice beer / coffee / whatever once this is merged 😉 🙏

leongersen commented:
I'll try to finish this up next week :)

leongersen commented:
Alright, I've added tests showing the new behavior and rebased against 5.4. I think I've tricked the pipeline into running the tests using the patched PhpUnit bridge for this PR, but I'm not sure if this is the correct way to do this. It took me a while to figure out how to get the tests to run using the changed bridge locally, too.

leongersen commented:
@nicolas-grekas / @ruudk  Sorry, I don't mean to nudge, but I'm not sure what the procedure is to report that this is ready for review. I don't think the remaining CI failure is caused by the changes in this PR.

fancyweb commented on src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler/Deprecation.php:
```diff
@@ -59,6 +60,10 @@ class Deprecation
      */
     public function __construct($message, array $trace, $file)
     {
+        if (isset($trace[2]['class']) && DebugClassLoader::class === $trace[2]['class']) {
```
Shouldn't we do it for `LegacyDebugClassLoader` (old class from symfony/debug) too?

leongersen replied:
Yes, I think so.

ruudk commented:
> @nicolas-grekas / @ruudk Sorry, I don't mean to nudge, but I'm not sure what the procedure is to report that this is ready for review. I don't think the remaining CI failure is caused by the changes in this PR.

I don't know either, but maybe @xabbuh or @ogizanagi can help getting this reviewed? 🙏

ruudk commented:
@leongersen could you rebase this PR so that we can get this merged?

leongersen commented:
Yes, done.

ruudk commented:
The Psalm job seems to be hanging while running the setup-php action. 
@leongersen You could try to do another force push to re-trigger it.

The failure in appveyor is related to `RedisExtIntegrationTest` and not caused by this PR.

ruudk commented:
@leongersen Oke, the Psalm errors need to be fixed. Are you able to do it? Or do you rather let me fix it on your branch?

stof commented:
Regarding the reported Psalm error, the one about `LegacyDebugClassLoader` not being defined is OK (as this is about supporting older Symfony version).
For DeprecationGroup not being defined, this look really weird.

leongersen commented:
I don't know what causes the Psalm error on `DeprecationGroup`. It was already `use`d in all classes before this PR.

ruudk commented:
I tried to run Psalm locally, but that is extremely tedious to do. There is no `psalm.phar` nor an installer instruction. The baseline seems to be generated on the fly in the CI, before running the code from the branch. Why was this done like this? Wouldn't it be easier if the baseline was part of the repository, and that there was a `psalm` script that can be invoked locally?

ruudk commented:
@nicolas-grekas Could you review this PR? Do you have an idea what could cause the Psalm error?

stof commented:
@ruudk generating the baseline on the fly is because we want each PR to report only the new things. The codebase is too far from using Psalm for everything to bother about maintaining a committed baseline (we don't always fix all things reported by psalm, depending on what it reports). This CI job is here as an helper, not as a requirement.

ruudk commented:
It would be really great if we could proceed with this PR. @leongersen sorry to ask, but could you update this PR one more time?

leongersen commented:
@ruudk, I've rebased on 5.4. Based on the comments by @stof, I'm not taking action on the Psalm failures.

nicolas-grekas commented:
Thank you @leongersen.

ruudk commented:
@leongersen @nicolas-grekas just tested 6.3.2 and it works! I'm so happy to finally to be able to hide the indirect Monolog deprecation 🥳🥳🥳 thanks for your work on this
